### PR TITLE
Gekidou - Edit Profile Screen

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
This PR caters for the 'Edit Profile' sub feature of the main [PR for the Account Screen](https://github.com/mattermost/mattermost-mobile/pull/5708)

However, we have already merged that main PR into the 'gekidou' branch, hence, this one will go into 'gekidou'.

```
Please note that it does not cater for the ProfileImage at the moment; i.e a user will not be able to change his profile image.

We will integrate this component later down the road once we are done with the Account screen.
```

#### Release Notes
```release-note
NONE
```

### CODE REVIEW

File: app/components/widgets/text_settings/components/label_content.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import {IntlText} from '@components/widgets/text_settings';
+import {useTheme} from '@context/theme';
+import {makeStyleSheetFromTheme} from '@utils/theme';
+
+type LabelContentProps = {
+    label: IntlText | string;
+    testID: string;
+};
+
+const LabelContent = ({label, testID}: LabelContentProps) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/components/label_content.tsx]: I refactored this area.

File: app/components/profile_picture/index.tsx [hunk 2]
```diff
@@ -22,12 +23,17 @@ const STATUS_BUFFER = Platform.select({
...
+    imageUri?: string;
+    profileImageRemove?: boolean;
+    profileImageUri?: string;
     showStatus?: boolean;
-    size: number;
+    size?: number;
+    status?: string;
     statusSize?: number;
-    statusStyle?: StyleProp<ViewProps>;
+    statusStyle?: StyleProp<ViewProps> | any;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: I have refactored this area.

File: app/components/profile_picture/index.tsx [hunk 3]
```diff
@@ -58,17 +64,20 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
     };
 });
 
-const ProfilePicture = ({author, iconSize, showStatus = true, size = 64, statusSize = 14, statusStyle, testID}: ProfilePictureProps) => {
+const ProfilePicture = ({author, showStatus = true, edit = false, iconSize, imageUri, profileImageRemove, profileImageUri, size = 128, status, statusSize = 14, statusStyle: defaultStatusStyle, testID}: ProfilePictureProps) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: I have refactored this area.

File: app/components/your_picture/index.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            borderWidth: 1,
+            borderColor: theme.centerChannelBg,
+        },
+        status: {
+            color: theme.centerChannelBg,
+        },
+    };
+});
+
+const YourPicture = ({author, iconSize, showStatus = true, size = 64, statusSize = 14, statusStyle, testID}: ProfilePictureProps) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/your_picture/index.tsx]: I have refactored this area.

File: app/screens/edit_profile/index.tsx [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        update[id] = name;
+        setUserInfo(update);
+
+        // @ts-expect-error access object property by string key
+        const currentValue = currentUser[id];
+        const didChange = currentValue !== name;
+        enableSaveButton(didChange);
+    }, [userInfo]);
+
+    const setScrollViewRef = useCallback((ref) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/edit_profile/index.tsx:230]: Yes, as the `KeyboardAwareScrollView` component has a prop called `innerRef` and this one expects a function.

File: app/screens/edit_profile/index.tsx [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+                setError(error);
+                setUpdating(false);
+                enableSaveButton(true);
+                scrollViewRef.current?.scrollToPosition(0, 0, true);
+                return;
+            }
+
+            close();
+        } catch (e) {
+            // console.error(e);
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/edit_profile/index.tsx:225]: Good catch - I'll do that

File: app/screens/edit_profile/index.tsx [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+    const service = currentUser.authService;
+
+    const rightButton = {
+        id: 'update-profile',
+        enabled: false,
+        showAsAction: 'always',
+        testID: 'edit_profile.save.button',
+        color: theme.sidebarHeaderTextColor,
+        text: intl.formatMessage({id: t('mobile.account.settings.save'), defaultMessage: 'Save'}),
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/edit_profile/index.tsx:146]: Nice catch - I will handle it

File: app/constants/files.ts [hunk 8]
```diff
@@ -13,6 +13,7 @@ const Files: Record<string, string[]> = {
     VIDEO_TYPES: ['mp4', 'avi', 'webm', 'mkv', 'wmv', 'mpg', 'mov', 'flv'],
     WORD_TYPES: ['doc', 'docx'],
     ZIP_TYPES: ['zip'],
+    VALID_MIME_TYPES: ['application/bmp', 'application/jpg', 'application/png', 'application/x-bmp', 'application/x-jpg', 'application/x-png', 'application/x-win-bitmap', 'image/bmp', 'image/jp_', 'image/jpeg', 'image/jpeg', 'image/jpg', 'image/ms-bmp', 'image/pipeg', 'image/pjpeg', 'image/png', 'image/vnd.swiftview-jpeg', 'image/x-bitmap', 'image/x-bmp', 'image/x-ms-bmp', 'image/x-win-bitmap', 'image/x-windows-bmp', 'image/x-xbitmap', 'image/x-xbitmap'],
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/constants/files.ts:16]: Done

File: app/components/widgets/text_settings/components/optional_content.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+type OptionalContentProps = {
+    optional: boolean;
+}
+
+const OptionalContent = ({optional}: OptionalContentProps) => {
+    const theme = useTheme();
+    const style = getStyleSheet(theme);
+
+    if (optional) {
+        <FormattedText
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/components/optional_content.tsx]: Removed

File: app/components/profile_picture_picker/index.tsx [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+
+    const prepareFileUpload = async (files: File[]) => {
+        const file = files?.[0];
+
+        if (!file) {
+            return;
+        }
+
+        if (!file.fileSize || !file.fileName) {
+            const path = Platform.OS === 'ios' ? (file?.path || file.uri).replace('file://', '') : (file?.path || file.uri);
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Corrected

File: app/components/post_list/post/avatar/index.tsx [hunk 11]
```diff
@@ -138,7 +138,7 @@ const Avatar = ({author, enablePostIconOverride, isAutoReponse, isSystemPost, pe
     });
 
     let component = (
-        <ProfilePicture
+        <YourPicture
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/post_list/post/avatar/index.tsx]: I have refactored this section

File: app/components/profile_picture/index.tsx [hunk 12]
```diff
@@ -77,31 +86,91 @@ const ProfilePicture = ({author, iconSize, showStatus = true, size = 64, statusS
...
     let image;
-    if (author && client) {
-        const pictureUrl = client.getProfilePictureUrl(author.id, author.lastPictureUpdate);
+    if (pictureUrl) {
+        let prefix = '';
+        if (Platform.OS === 'android' && !pictureUrl.startsWith('content://') && !pictureUrl.startsWith('http://') && !pictureUrl.startsWith('https://')) {
+            prefix = 'file://';
+        }
+
+        if (pictureUrl.includes('api')) {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: Modified so that it searches for '/api/' instead of 'api' only.

File: app/components/widgets/text_settings/index.tsx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    defaultMessage: string;
+}
+
+type TextSettingProps = {
+    disabled?: boolean;
+    disabledText?: string;
+    errorText?: string;
+    helpText?: string | number;
+    id: string;
+    keyboardType?: 'default' | 'number-pad' | 'decimal-pad' | 'numeric' | 'email-address' | 'phone-pad' | 'url';
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/index.tsx]: This component has been removed.

File: app/screens/navigation.ts [hunk 14]
```diff
@@ -359,12 +359,12 @@ export function showSearchModal(initialValue = '') {
     showModal(name, title, passProps, options);
 }
 
-export async function dismissModal(options = {}) {
+export async function dismissModal(options: any = {}) {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/navigation.ts]: Done.

File: app/components/profile_picture_picker/index.tsx [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+            android: 36,
+        }),
+        width: 45,
+        alignItems: 'center',
+        justifyContent: 'center',
+    },
+});
+
+const ProfilePictureButton = ({
+    browseFileTypes = Platform.OS === 'ios' ? 'public.item' : '*/*',
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Done.

File: app/components/profile_picture_picker/index.tsx [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+    uploadFiles,
+    wrapper,
+}: ProfilePictureButtonProps) => {
+    const intl = useIntl();
+    const theme = useTheme();
+    const serverUrl = useServerUrl();
+    const getPermissionDeniedMessage = (source: string) => {
+        switch (source) {
+            case 'camera': {
+                return getPermissionMessages(intl).camera;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Refactored

File: app/components/profile_picture_picker/index.tsx [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+    onShowFileSizeWarning,
+    onShowUnsupportedMimeTypeWarning,
+    removeProfileImage,
+    uploadFiles,
+    wrapper,
+}: ProfilePictureButtonProps) => {
+    const intl = useIntl();
+    const theme = useTheme();
+    const serverUrl = useServerUrl();
+    const getPermissionDeniedMessage = (source: string) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Refactored

File: app/components/widgets/text_settings/index.tsx [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    optional = false,
+    secureTextEntry = false,
+    testID,
+    value,
+    onBlur,
+}: TextSettingProps) => {
+    const theme = useTheme();
+    const intl = useIntl();
+
+    const onChangeText = useCallback((text: string) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/index.tsx]: Refactored

File: app/components/widgets/text_settings/index.tsx [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        return onChange(id, text);
+    }, []);
+
+    const onBlurField = useCallback(() => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/index.tsx]: Refactored

File: app/components/widgets/text_settings/index.tsx [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        return onBlur?.(id);
+    }, []);
+
+    const style = getStyleSheet(theme);
+    const textStyles = getMarkdownTextStyles(theme);
+    const blockStyles = getMarkdownBlockStyles(theme);
+
+    const keyboard = Platform.OS === 'android' && keyboardType === 'url' ? 'default' : keyboardType;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/widgets/text_settings/index.tsx]: I will keep this one as it's equivalent to the below but shorter: 

```
   let keyboard = keyboardType;

    if (Platform.OS === 'android' && keyboardType === 'url') {
        keyboard = 'default';
    }
```

File: app/components/profile_picture_picker/index.tsx [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+            if (uri) {
+                files.push({...file, fileName, uri, type});
+            }
+        } else {
+            files.push(file);
+        }
+        return files;
+    };
+
+    const hasPhotoPermission = async (source: string) => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Refactored

File: app/components/profile_picture_picker/index.tsx [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+                if (Platform.OS === 'android') {
+                    return getPermissionMessages(intl).denied.android;
+                }
+
+                return getPermissionMessages(intl).denied.ios;
+            }
+        }
+    };
+
+    const attachFileFromCamera = async () => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Done

File: app/components/profile_picture_picker/index.tsx [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+                res.uri = decodeURIComponent(res.uri);
+
+                prepareFileUpload([res]);
+            } catch (error) {
+                // Do nothing
+            }
+        }
+    };
+
+    const getFilesFromResponse = async (response: FileResponse): Promise<File[]> => {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Done

File: app/components/profile_picture_picker/index.tsx [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+                                defaultMessage: "Don't Allow",
+                            }),
+                        },
+                    ]);
+                    return false;
+                }
+            }
+        }
+
+        return true;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: Do you mean to return immediately for the platform that does have the permission by default?

File: app/components/profile_picture_picker/index.tsx [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+
+        try {
+            client = NetworkManager.getClient(serverUrl);
+            profileImageUrl = client.getProfilePictureUrl(id, lastPictureUpdate);
+        } catch {
+            // does nothing
+        }
+
+        // Check if image url includes query string for timestamp. If so, it means the image has been updated from the default, i.e. '.../image?_=1544159746868'
+        if (profileImageUrl?.includes('?')) {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: We are verifying the result from `getProfilePictureUrl` method of the rest client.  If it does contain `?`, then it would mean that the user has an existing profile image on the server.  Am I missing anything?

File: app/components/profile_picture_picker/index.tsx [hunk 26]
```diff
@@ -0,0 +1,454 @@
...
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    };
+
+    const hasStoragePermission = async () => {
+        if (Platform.OS === 'android') {
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: @larkox Do you mean to return immediately for the platform that does have the permission by default?

File: app/components/profile_picture/index.tsx [hunk 27]
```diff
@@ -77,31 +86,91 @@ const ProfilePicture = ({author, iconSize, showStatus = true, size = 64, statusS
...
+            uri: `${prefix}${pictureUrl}`,
+        };
+
         image = (
             <FastImage
                 key={pictureUrl}
-                style={{width: size, height: size, borderRadius: (size / 2)}}
-                source={{uri: `${serverUrl}${pictureUrl}`}}
+                style={{width: size, height: size, borderRadius: size / 2}}
+                source={source}
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: In the refactored code, I'm using memo on the 'image' component.  Yes, the FastImage does [cache images](https://github.com/DylanVann/react-native-fast-image/blob/main/docs/how-is-caching-handled.md).

File: app/actions/remote/user.ts [hunk 28]
```diff
@@ -451,3 +451,21 @@ export const unsetCustomStatus = async (serverUrl: string) => {
...
+    let client: Client;
+
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    try {
+        await client.setDefaultProfileImage(userId);
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/actions/remote/user.ts:466]: I refactored the code for it to handle the `last_picture_update`.

File: app/components/profile_picture_picker/index.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (larkox) [reply app/components/profile_picture_picker/index.tsx]: I am trying to think of any other weird situation where a `?` may appear in that url (like a `?` in the base url? (that should not be possible as far as I know)). Maybe I am worrying too much and this is fine as it is.

[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture_picker/index.tsx]: I can check for `image?_` instead of only `?`

File: app/components/profile_picture/index.tsx [hunk 29]
```diff
@@ -77,31 +86,91 @@ const ProfilePicture = ({author, iconSize, showStatus = true, size = 64, statusS
...
+        if (edit && imageUri) {
+            setPictureUrl(imageUri);
+        }
+    }, [edit, imageUri]);
+
+    useDidUpdate(() => {
+        if (profileImageUri) {
+            setPictureUrl(profileImageUri);
+        }
+    }, [profileImageUri]);
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: I have clean up this part.  ProfileImageURI was not being used and thus I removed it.

File: app/components/profile_picture/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/components/profile_picture/index.tsx]: Corrected.

File: app/actions/local/user.ts [hunk 30]
```diff
@@ -105,25 +105,25 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
-        try {
-            await operator.batchRecords([user]);
-        } catch {
-            // eslint-disable-next-line no-console
-            console.log('FAILED TO BATCH CHANGES FOR UPDATE USER PRESENCE');
-        }
+    } catch {
+        // eslint-disable-next-line no-console
+        console.log('FAILED TO BATCH CHANGES FOR UPDATE USER PRESENCE');
+        return {};
```
Comment (enahum) [inline app/actions/local/user.ts:125]: log seems wrong as you are not batching any more and in case it fails should't you return the error?

File: app/actions/local/user.ts [hunk 31]
```diff
@@ -105,25 +105,25 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
-    const user = await queryUserById(operator.database, userStatus.user_id);
-    if (user) {
-        user.prepareUpdate((record) => {
-            record.status = userStatus.status;
+    try {
+        const user = await database.get(MM_TABLES.SERVER.USER).find(userId) as UserModel;
+        await database.write(async () => {
+            await user.update((userRecord: UserModel) => {
+                userRecord.status = userDetails?.status ?? user.status;
+                userRecord.lastPictureUpdate = userDetails?.lastPictureUpdate ?? user.lastPictureUpdate;
```
Comment (enahum) [inline app/actions/local/user.ts:119]: so you replaced the `updateUserPresence` function for something more generic which is fine, but here you are only only updating two values instead of every possible property. it seems to me that this needs more work or you keep them separately.

Also the problem you'll face here is that the argument is a `Partial<UserModel>` basically forcing you to initialize a UserModel in other places as you cannot set it's values within an existing object directly. So in my opinion to avoid that from happening you'll need to pass `Partial<UserProfile>` instead, that way you can quickly create an UserProfile object anywhere and easily.

File: app/actions/remote/user.ts [hunk 32]
```diff
@@ -383,13 +384,13 @@ export const setStatus = async (serverUrl: string, status: UserStatus) => {
 
     try {
         const data = await client.updateStatus(status);
-        updateUserPresence(serverUrl, status);
+        await updateLocalUser(serverUrl, status.user_id, {status: status.status});
```
Comment (enahum) [inline app/actions/remote/user.ts:387]: this makes my previous point

File: app/actions/remote/user.ts [hunk 33]
```diff
@@ -451,3 +452,22 @@ export const unsetCustomStatus = async (serverUrl: string) => {
...
+
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    try {
+        await client.setDefaultProfileImage(userId);
+        updateLocalUser(serverUrl, userId, {lastPictureUpdate: 0});
```
Comment (enahum) [inline app/actions/remote/user.ts:467]: this should not be 0, you want to set the Date.now() here

File: app/components/floating_text_input_label/animated_input/index.tsx [hunk 34]
```diff
@@ -238,6 +238,7 @@ const FloatingTextInput = forwardRef<FloatingTextInputRef, FloatingTextInputProp
                     </Animated.Text>
                 }
                 <TextInput
+                    autoCapitalize={'none'}
```
Comment (enahum) [inline app/components/floating_text_input_label/animated_input/index.tsx:241]: if you want this to apply to all of them, that you shouldn't want that, this prop should go below `{...props}` the reality is that you should pass this as a prop on the parent component.

File: app/components/post_list/post/avatar/index.tsx [hunk 35]
```diff
@@ -11,7 +11,7 @@ import {of as of$} from 'rxjs';
 import {switchMap} from 'rxjs/operators';
 
 import CompassIcon from '@components/compass_icon';
-import ProfilePicture from '@components/profile_picture';
+import UserProfilePicture from '@components/profile_picture/image';
```
Comment (enahum) [inline app/components/post_list/post/avatar/index.tsx:14]: this is throwing me off.. should it be something like
```suggestion
import ProfileImage from '@components/profile_picture/image';
```

File: app/components/profile_picture/edit_image/index.tsx [hunk 36]
```diff
@@ -0,0 +1,151 @@
...
+
+import {Client} from '@client/rest';
+import CompassIcon from '@components/compass_icon';
+import {useServerUrl} from '@context/server_url';
+import {useTheme} from '@context/theme';
+import useDidUpdate from '@hooks/did_update';
+import NetworkManager from '@init/network_manager';
+import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
+
+import UserProfileImage from '../image';
```
Comment (enahum) [inline app/components/profile_picture/edit_image/index.tsx:15]: hmmm, feels wrong

File: app/components/profile_picture/image/index.tsx [hunk 37]
```diff
@@ -0,0 +1,88 @@
...
+            color: changeOpacity(theme.centerChannelColor, 0.48),
+        },
+    };
+});
+
+type UserProfileImageProps = {
+    imageUrl?: string;
+    lastPictureUpdate: number;
+    size: number;
+    source?: { uri: string };
```
Comment (enahum) [inline app/components/profile_picture/image/index.tsx:33]:
```suggestion
    source?: {uri: string};
```

File: app/components/profile_picture/edit_image/index.tsx [hunk 38]
```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline app/components/profile_picture/edit_image/index.tsx:1]: not sure why this component is under `profile_picture` they seem unrelated to me

File: app/components/profile_picture/edit_image/index.tsx [hunk 39]
```diff
@@ -0,0 +1,151 @@
...
+    testID?: string;
+    userId: string;
+};
+
+const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
+    return {
+        container: {
+            justifyContent: 'center',
+            alignItems: 'center',
+            borderRadius: 80,
```
Comment (enahum) [inline app/components/profile_picture/edit_image/index.tsx:32]: why is this radius 80? doesn't this has direct relation with the size prop ?

File: app/components/profile_picture/edit_image/index.tsx [hunk 40]
```diff
@@ -0,0 +1,151 @@
...
+            borderColor: theme.centerChannelBg,
+            borderRadius: 36 / 2,
+            height: 36,
+            width: 36,
+        },
+    };
+});
+
+const EditProfilePicture = ({
+    canUpdateProfilePicture = false,
```
Comment (enahum) [inline app/components/profile_picture/edit_image/index.tsx:52]: what is the reason behind this ?

File: app/components/profile_picture/edit_image/index.tsx [hunk 41]
```diff
@@ -0,0 +1,151 @@
...
+    imageUri,
+    isProfileImageRemoved,
```
Comment (enahum) [inline app/components/profile_picture/edit_image/index.tsx:54]: why is this a prop?

**Review by enahum — CHANGES_REQUESTED** [review]
partially reviewed..

File: app/actions/local/user.ts (hunk 31 above)
Comment (avinashlng1080) [reply app/actions/local/user.ts:119]: Thanks Elias.  I wanted to have something similar to how the Operator's transformer were mapping the field.  I have updated this code now for it to include the other fields as well.

File: app/actions/remote/user.ts (hunk 32 above)
Comment (avinashlng1080) [reply app/actions/remote/user.ts:387]: Noted and corrected

File: app/actions/remote/user.ts (hunk 33 above)
Comment (avinashlng1080) [reply app/actions/remote/user.ts:467]: Corrected

File: app/components/profile_picture/image/index.tsx (hunk 37 above)
Comment (avinashlng1080) [reply app/components/profile_picture/image/index.tsx:33]: corrected

File: app/components/profile_picture/edit_image/index.tsx (hunk 40 above)
Comment (avinashlng1080) [reply app/components/profile_picture/edit_image/index.tsx]: removed

File: app/components/profile_picture/edit_image/index.tsx (hunk 39 above)
Comment (avinashlng1080) [reply app/components/profile_picture/edit_image/index.tsx:31]: corrected

File: app/screens/edit_profile/components/field.tsx [hunk 42]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {useIntl} from 'react-intl';
```
Comment (avinashlng1080) [inline app/screens/edit_profile/components/field.tsx:5]: This component should be removed in favor of the InputField component

File: app/actions/local/user.ts [hunk 43]
```diff
@@ -105,25 +105,39 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
     });
 };
 
-export const updateUserPresence = async (serverUrl: string, userStatus: UserStatus) => {
-    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
-    if (!operator) {
+export const updateLocalUser = async (serverUrl: string, userId: string, userDetails: Partial<UserProfile> & { status?: string}) => {
```
Comment (enahum) [inline app/actions/local/user.ts:108]: you don't need `userId` as is part of `userDetails`, just make sure the id is set before doing anything else

File: app/actions/local/user.ts [hunk 44]
```diff
@@ -105,25 +105,39 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
-    if (user) {
-        user.prepareUpdate((record) => {
-            record.status = userStatus.status;
+    try {
+        const user = await database.get(MM_TABLES.SERVER.USER).find(userId) as UserModel;
+        await database.write(async () => {
+            await user.update((userRecord: UserModel) => {
+                userRecord.status = userDetails?.status ?? user.status;
+                userRecord.authService = userDetails.auth_service ?? user.authService;
+                userRecord.deleteAt = userDetails.delete_at ?? user.deleteAt;
```
Comment (enahum) [inline app/actions/local/user.ts:120]: you are not really deleting the user locally like this unless the server tells you to delete it, basically this will never happen here

File: app/actions/local/user.ts [hunk 45]
```diff
@@ -105,25 +105,39 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
+                userRecord.email = userDetails.email ?? user.email;
+                userRecord.firstName = userDetails.first_name ?? user.firstName;
+                userRecord.isBot = userDetails.is_bot ?? user.isBot;
+                userRecord.isGuest = userDetails?.roles?.includes('system_guest') ?? user.isGuest;
```
Comment (enahum) [inline app/actions/local/user.ts:124]: same as delete_at

File: app/actions/local/user.ts [hunk 46]
```diff
@@ -105,25 +105,39 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
+                userRecord.lastName = userDetails.last_name ?? user.lastName;
+                userRecord.lastPictureUpdate = userDetails.last_picture_update ?? user.lastPictureUpdate;
+                userRecord.locale = userDetails.locale ?? user.locale;
+                userRecord.nickname = userDetails.nickname ?? user.nickname;
+                userRecord.notifyProps = userDetails.notify_props ?? user.notifyProps;
+                userRecord.position = userDetails?.position ?? '' ?? user.position;
```
Comment (enahum) [inline app/actions/local/user.ts:130]:
```suggestion
                userRecord.position = userDetails?.position ?? user.position;
```

File: app/actions/local/user.ts [hunk 47]
```diff
@@ -105,25 +105,39 @@ export const updateRecentCustomStatuses = async (serverUrl: string, customStatus
...
+                userRecord.props = userDetails.props ?? user.props;
+                userRecord.roles = userDetails.roles ?? user.roles;
+                userRecord.timezone = userDetails.timezone ?? user.timezone;
+                userRecord.updateAt = userDetails.update_at ?? user.updateAt;
```
Comment (enahum) [inline app/actions/local/user.ts:134]: this is not updated client side

File: app/components/post_list/post/avatar/index.tsx [hunk 48]
```diff
@@ -139,10 +139,9 @@ const Avatar = ({author, enablePostIconOverride, isAutoReponse, isSystemPost, pe
 
     let component = (
         <ProfilePicture
-            author={author}
+            lastPictureUpdate={author.lastPictureUpdate}
```
Comment (enahum) [inline app/components/post_list/post/avatar/index.tsx:142]: revert this change as you need to display the status for posts

**Review by enahum — CHANGES_REQUESTED** [review]
re-refactor needed

File: app/components/file_picker/index.tsx [hunk 49]
```diff
@@ -0,0 +1,142 @@
...
+    userId,
+}: ProfilePictureButtonProps) => {
+    const intl = useIntl();
+    const theme = useTheme();
+    const serverUrl = useServerUrl();
+    const pictureUtils = useMemo(
+        () =>
+            new PickerUtil(
+                intl,
+                maxFileSize,
```
Comment (enahum) [inline app/components/file_picker/index.tsx:54]: I believe you need to include this in the memo deps

File: app/utils/file/picker_util.ts [hunk 50]
```diff
@@ -0,0 +1,360 @@
...
+        try {
+            client = NetworkManager.getClient(serverUrl);
+            profileImageUrl = client.getProfilePictureUrl(userId, lastPictureUpdate);
+        } catch {
+            return false;
+        }
+
+        // Check if image url includes query string for timestamp. If so, it means the image has been updated from the default, i.e. '.../image?_=1544159746868'
+        return Boolean(profileImageUrl?.includes('image?_'));
+    };
```
Comment (enahum) [inline app/utils/file/picker_util.ts:359]: huh? what is going on here? `client/getProfilePictureUrl` is a local function that constructs the url, this will ALWAYS be true!!!

File: app/components/profile_picture/index.tsx [hunk 51]
```diff
@@ -14,120 +14,132 @@ import NetworkManager from '@init/network_manager';
...
-
-type ProfilePictureProps = {
-    author?: UserModel;
-    iconSize?: number;
-    showStatus?: boolean;
-    size: number;
-    statusSize?: number;
-    statusStyle?: StyleProp<ViewProps>;
-    testID?: string;
-};
```
Comment (enahum) [inline app/components/profile_picture/index.tsx:32]: why is this component being modified?? Do revert

File: app/screens/edit_profile/edit_profile.tsx [hunk 52]
```diff
@@ -0,0 +1,407 @@
...
+
+    const submitUser = async () => {
+        enableSaveButton(false);
+        setError(undefined);
+        setUpdating(true);
+        try {
+            if (profileImage) {
+                const now = Date.now();
+                await uploadProfileImage();
+                updateLocalUser(serverUrl, {last_picture_update: now});
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.tsx:194]: if the image is removed, last_picture_update should be equal to 0

**Review by enahum — CHANGES_REQUESTED** [review]
the issues we've discussed before has not been address yet.. wondering if you should clean up the entire branch, 263 commits seems too much for this. Let me know if you need help and I'll take this from you

**Comment by enahum** [comment]
closing this PR as it needs to be cleaned and it already is too confusing.
